Allow user specify max items per page

mcsky opened the pull request:
- Allow configuration to let user specify the limit of items per page

- Add more informations about pagination data to collection response

Commit:
Allow configuration to let user specify the limit of items per page
Allow configuration of a max number of items retrievable by route. This limit can't be surpassed
Add more informations about pagination data to collection response

Nek- commented on src/Crud/CrudControllerInterface.php:
```diff
@@ -29,4 +29,12 @@ interface CrudControllerInterface
+     *
+     * Otherwise ignore "size" key
+     */
+    public const TAKE_USER_MAX_PAGE = 'melodiia_take_user_max_per_page';
```
ALLOW_USER_DEFINED

Nek- commented on src/Crud/Controller/GetAll.php:
```diff
@@ -62,6 +62,11 @@ public function __invoke(Request $request)
         $page = $request->query->getInt('page', 1);
         $maxPerPage = $request->attributes->get(self::MAX_PER_PAGE_ATTRIBUTE, 30);
+        if (true === $request->attributes->getBoolean(self::TAKE_USER_MAX_PAGE, false)) {
+            if (0 !== $userMax = $request->query->getInt('size', 0)) {
```
`item_per_page` would be a better name. Having it as const would be also better. Finally, we need some more configuration considering we can't allow the user to get the whole database. We also need a `maximum_items_per_page` config (that could be 300 by default).

Nek- commented on src/Serialization/Json/OkContentNormalizer.php:
```diff
@@ -56,7 +56,12 @@ public function normalize($object, $format = null, array $context = array())
+                'countPages'  => $content->getNbPages(),
+                'countResult' => $content->getNbResults(),
+                'currentPage' => $content->getCurrentPage(),
+                'maxPerPage'  => $content->getMaxPerPage(),
```
This is out of apijson specs. But makes a lot of sense. 👍

Nek- replied:
But `totalPages` and `totalResult` aren't better names?

Here are pros:
1. Keeping `totalPages` does not breaks compatibility
2. [Hydra](https://github.com/HydraCG/Specifications/blob/master/drafts/use-cases/3.2.pagination.md) also uses `total` wording for this kind of stuff

mcsky replied:
Yes totally my bad, I just wrote that on my current project, so I didn't take care about BC. ^^'

Nek- reviewed (changes requested):
Thanks for this contribution!

It looks great but is missing some things (and security). I highlighted things not ok to me. Feel free to challenge my notices.

Nek- commented on src/Crud/Controller/GetAll.php:
```diff
@@ -60,8 +60,18 @@ public function __invoke(Request $request)
-        $page = $request->query->getInt('page', 1);
+        $page       = $request->query->getInt('page', 1);
         $maxPerPage = $request->attributes->get(self::MAX_PER_PAGE_ATTRIBUTE, 30);
+        if (true === $request->attributes->getBoolean(self::ALLOW_USER_DEFINED_MAX_PAGE, false)) {
```
It's a `getBoolean` so you really don't need to check against true. Please transform to the following:

```php
if ($request->attributes->getBoolean(self::ALLOW_USER_DEFINED_MAX_PAGE, false))
```

Nek- commented on src/Crud/PagesRequestFactory.php:
```diff
@@ -0,0 +1,31 @@
+        $page       = $request->query->getInt('page', 1);
+        $maxPerPage = $request->attributes->get(CrudControllerInterface::MAX_PER_PAGE_ATTRIBUTE, 30);
+        if (true === $request->attributes->getBoolean(CrudControllerInterface::ALLOW_USER_DEFINED_MAX_PAGE, false)) {
+            $maxPerPageQueryAttribute = $request->attributes->get(CrudControllerInterface::MAX_PER_PAGE_QUERY_ATTRIBUTE, 'max_per_page');
```
This configuration by controller disturbs me a lot. It should really be in global configuration. (I mean, having it by controller is a mistake to me)

This implies to use the configuration object of Melodiia for cruds...

mcsky replied:
Hummm it could be configured in melodiia global configuration. But in some cases, you may want to increase this limit for one endpoint (example: leads).

Nek- replied:
It's not about the size, it's about the parameter name!

mcsky replied:
Under which key in configuration, do you want it?

Nek- commented on src/Crud/PagesRequestFactory.php:
```diff
@@ -0,0 +1,31 @@
+    {
+        $page       = $request->query->getInt('page', 1);
+        $maxPerPage = $request->attributes->get(CrudControllerInterface::MAX_PER_PAGE_ATTRIBUTE, 30);
+        if (true === $request->attributes->getBoolean(CrudControllerInterface::ALLOW_USER_DEFINED_MAX_PAGE, false)) {
```
```php
if (true === $request->attributes->getBoolean(CrudControllerInterface::ALLOW_USER_DEFINED_MAX_PAGE, false)
```

Nek- commented on src/Crud/PagesRequest.php:
```diff
@@ -0,0 +1,35 @@
+
+namespace Biig\Melodiia\Crud;
+
+class PagesRequest
```
A new subnamespace `Pagination` looks accurate. (because pagination is a real thing)

mcsky replied:
I hesitate, then I saw no namespace for filters (and it's a real thing too 👼)

Nek- commented on src/Crud/PagesRequestFactory.php:
```diff
@@ -0,0 +1,31 @@
+
+use Symfony\Component\HttpFoundation\Request;
+
+class PagesRequestFactory
```
This class is definitely a goog idea, but using static method without interface, you enforce its usage and make it useless.
Please make it service with an interface so users can define another one.

mcsky replied:
Done

need to add user service auto-registration if the interface is implemented

Nek- replied:
I don't get your point. I don't think this is required. If needed the user just overrides the service.

Nek- commented on src/Crud/Persistence/DataStoreInterface.php:
```diff
@@ -18,12 +19,14 @@ public function save(object $model);
     public function find(string $type, $id): ?object;
 
     /**
+     *
```
seems useless

Nek- commented on src/Bridge/Doctrine/DoctrineDataStore.php:
```diff
@@ -43,8 +44,9 @@ public function getPaginated(string $type, int $page, FilterCollectionInterface
-        $pager->setMaxPerPage($maxPerPage);
+        // If pages request is specified
+        $pager->setCurrentPage($pagesRequest ? $pagesRequest->getPage() : $page);
+        $pager->setMaxPerPage($pagesRequest ? $pagesRequest->getMaxPerPage() :$maxPerPage);
```
The datastore should not handle anything out of its scope (generating a pager will be its limit). Please replace this code inside the controller.

Nek- commented on src/Bridge/Doctrine/DoctrineDataStore.php:
```diff
@@ -31,7 +32,7 @@ public function find(string $type, $id): ?object
     }
 
-    public function getPaginated(string $type, int $page, FilterCollectionInterface $filters, $maxPerPage = 30): PagerFanta
+    public function getPaginated(string $type, int $page, FilterCollectionInterface $filters, $maxPerPage = 30, PagesRequest $pagesRequest = null): PagerFanta
```
This is a useless BC break.

mcsky replied:
I added this against a BC actually... Why a nullable last parameter on an interface causes a fatal error? It's stupid to me right now

Nek- replied:
@mcsky 
It's not stupid, it's breaking because:
1. This bc break is useless because as I said the datastore should not alter the pager
2. It is a bc break because an interface is designed to be implemented (unless it's tagged `@internal`). See [it in action here](https://3v4l.org/c95AB).
3. Please be kind, I'm not that stupid :) (got it, you was saying the new parameter is stupid because it's not necessary for datastore, that's what I was saying by saying "useless")

Nek- reviewed (changes requested):
Some things are cool in your implementation, some others are really a problem. The modification of the datastore is the worth to me. It's responsibility is to return data, not to manage how they will be rendered. I was not sure about returning a paginator at the very first start, but it turned out that it's a great thing. But modifying this paginator inside the datastore is too much.

Nek- commented on src/Bridge/Symfony/Resources/config/crud.yaml:
```diff
@@ -10,6 +10,9 @@ services:
             $filters: !tagged melodiia.crud_filter
 
+    melodiia.crud.pagination.pages_request_factory:
+        class:
```
Something is missing over here :) ... And obviously, the change of the definition of the crud is missing too.

mcsky replied:
what do you mean by `change of the definition of the crud is missing too`

Nek- replied:
The current configuration is not complete according to your work:

```yaml
    melodiia.crud.controller.get_all:
        class: Biig\Melodiia\Crud\Controller\GetAll
        arguments:
            $dataStore: '@melodiia.doctrine.data_provider'
            $checker: '@security.authorization_checker'
            $collectionFactory: '@melodiia.crud.filters.filter_collection_factory'
        tags: [ controller.service_arguments ]
```

mcsky replied:
It's ok now, good catch 👍

Commit:
make PR changes
Allow possibility to configure the query parameter id, where melodiia will retrieve count of items asked by client
Add an overridable interface, that melodiia will use to do the line above
Update documentation
Add tests

mcsky commented:
The below has been add/update

- Allow possibility to configure the query parameter id, where melodiia will retrieve count of items asked by client
- Add an overridable interface, that melodiia will use to do the line above
- Update documentation

I'll add the tests soon 🔢

mcsky commented:
@Nek- Tests added, this MR is ready to review for me :)

Nek- commented on src/Bridge/Symfony/Resources/config/crud.yaml:
```diff
@@ -34,6 +42,7 @@ services:
             $dataStore: '@melodiia.doctrine.data_provider'
             $checker: '@security.authorization_checker'
             $collectionFactory: '@melodiia.crud.filters.filter_collection_factory'
+            $pagesRequestFactory: '@Biig\Melodiia\Crud\Pagination\PaginationRequestFactoryInterface'
```
`@melodiia.crud.pagination.pages_request_factory`

Nek- commented on docs/Symfony.md:
````diff
@@ -43,10 +43,25 @@ melodiia:
+```yaml
+#services.yaml
+
+Biig\Melodiia\Crud\Pagination\PaginationRequestFactoryInterface:
````
`melodiia.crud.pagination.pages_request_factory`

mcsky replied:
Actually the FQCN of this interface is bound to the service `melodiia.crud.pagination.pages_request_factory` by default.

The user can override this binding, and map the interface to its own class.

This is the wanted behavior :)

Nek- commented on docs/Symfony.md:
````diff
@@ -43,10 +43,25 @@ melodiia:
             datetime: true
 ```
+
+III. Override Melodiia services
````
Note: this is quite obvious when you use Symfony. But well nothing wrong to say it here to me. (some people would say we should not paraphrase Symfony docs)

Nek- commented on docs/Symfony.md:
```diff
@@ -43,10 +43,25 @@ melodiia:
                 This is an awesome API. Really.
             version: 1.0.0
+            pagination:
+                max_per_page_attribute: 'FooQueryParameter'
```
This is not the default value (you should write the default value here). Which is `max_per_page`.

Nek- commented on src/MelodiiaConfigurationInterface.php:
```diff
@@ -21,4 +24,21 @@ public function getDocumentationConfig(): array;
+     *
+     * @return Route[]
+     */
+    public function getRoutes(): array;
```
Cette méthode n'est pas liée à la configuration directement, et je ne crois pas qu'elle soit utile, il faudrait la passer private et la supprimer de l'interface !

Nek- reviewed (changes requested):
This looks really nice. :) 

But I'm sorry, I still have some part of the code that looks not ok to me (mostly this interface that really disturb me). The wrong service name usage is also a problem.

Furthermore, I see here some BC breaks:
1. ~~CrudControllerInterface~~
2. MelodiiaConfigurationInterface
3. ~~GetAll new constructor parameter~~

Please correct me if I'm wrong (I may have miss stuff). It's for changelog writing purpose.

mcsky commented:
MR updated according to your review.

Considering the following potential BCs: 
- CrudControllerInterface -> I just add new `const`, for me no BC here.
- MelodiiaConfigurationInterface -> The new method `getApiConfigFor` is a BC yes. But IMO nobody overrides MelodiiaConfiguration for now. Nevertheless, we should release this PR under a minor version, and notice this in the changelog.
- GetAll new constructor parameter
GetAll is a final class, so can't be inherited. So it will always up to date.
